### ISSUE TITLE
Error outputs on activities

### ISSUE THREAD

**Issue by tomy2105** [elsa-workflows/elsa-core#1115]
Splitting #1100....

Designer application writer must be able to handle (at least some of the errors) that happen during workflow execution.
**This is a must for any serious workflow application.**
Currently it cannot because error outputs on activities are missing, they have just done or workflow fault.

One obvious example would be network glitch when using http request, I want to be able to know and try to retry if I wish so from within designer. Currently it would just fault the workflow. This one is actually a showstopper because you cannot workaround it in any way.....

**Comment by sfmskywalker** [comment]
I think this is very specific to a given activity, because not all activities may cause transient errors.

For example, the **ForEach** activity will never fault unless you entered an invalid expression, which should fault the workflow.

But with the case of **Send HTTP Request**, that is indeed a good example of an activity that should allow the user to control what happens in case of a transient network error.

What we can do in these cases is wrap the call with try/catch, and return an outcome of e.g. "Network Error" or whatever is appropriate.

We should also allow the application developer to configure a retry policy to deal with transient errors. Only when even the retry policy gives up, will an exception be thrown and then be caught by the activity.

The right behavior needs to be determined per activity on a case-by-case basis.
Perhaps you want to split this out into various issues (one per activity), and keep this issue as an epic?

**Comment by tomy2105** [comment]
> I think this is very specific to a given activity, because not all activities may cause transient errors.
> The right behavior needs to be determined per activity on a case-by-case basis.

Indeed, some simple activities, Like **ForEach** won't have any outputs but Done, but anything else bit more complex probably will :).

> Perhaps you want to split this out into various issues (one per activity), and keep this issue as an epic?

I haven't looked at all activities that come with Elsa yet but I count something like 70.
Even if half need additional outputs it would produce 30ish tasks :(.
You really want one task per activity?

**Comment by sfmskywalker** [comment]
Good point. Maybe a simple checklist in this issue suffices 😅

**Comment by tomy2105** [comment]
I tried to keep the list minimal (but anything dealing with database/network is bound to be able to fail) :)...

Redirect 
SendHttpRequest 
WriteHttpResponse 
SendEmail 
SaveToDropbox 
ClearTimer 
Cron 
StartAt 
Timer 
PublishRebusMessage 
SendRebusMessage 
BlobExists 
DeleteBlob 
ReadBlob 
WriteBlob 
DeleteFile 
EnumerateFiles 
FileExists 
OutFile 
ReadFile 
TempFile

**Comment by sfmskywalker** [comment]
In addition, we need some sort of compensation mechanism in place that allows the workflow editor to execute cleanup logic other than in direct response to a faulting activity. For example, Activity A might create Azure resource, and at the very end of the workflow, Activity Z faults. Basically every activity needs a "compensation" outcome so that each previously executed step can be cleaned up

See: #2683

**Comment by angularsen** [comment]
Hi, I'm new and just started browsing Elsa docs and issues.

### HTTP Request resiliency
For the case of HTTP requests, is there any built-in resiliency? I could not see any.
We are looking for frameworks to help us build workflows connecting multiple HTTP APIs in a resilient manner.

Some default retry would go a long way, but making it configurable would be nice:
- Max retry count or max retry time period.
- Linear or exponential backoff.
- Timeout per retry or total timeout for all retries .

Currently we use [Polly](https://github.com/App-vNext/Polly) to make all our HTTP requests more resilient, but it is only short-lived retries for those transient network errors or hiccups on the receiving end. Something similar could be built into the HTTP request activity.

### Auto-retry on any activity
However, I just realized with the capability of Elsa and long-lived workflows, it might also be possible to retry for hours or even days before giving up and failing the workflow, in case of external downtime events. 

Maybe this is already supported, but is it not possible to build a retry capability for any activity? 
Perhaps using a Timer to periodically signal an activity to run until it succeeds, or gives up by some criteria.

The visual designer could show a short retry loop back to the same activity, blocking on the next retry, or have the activity be in a retry state instead of failed state.

This retry would then work on all kinds of activities that declare they may fail, including custom activities.
Each retry would hopefully be documented and tracked in the journal, and we could read information about each failure while investigating what is going on.

Just 2 cents from someone looking into Elsa for the first time, there is much to love about Elsa from what I've read so far! ❤️
